File: Fibonacci.hs
fibonacci :: Integer -> Integer
fibonacci 0 = 0
fibonacci 1 = 1
fibonacci n | n > 0 = fibonacci(n - 1) + fibonacci(n - 2)
            | n < 0 = fibonacci(n + 2) - fibonacci(n + 1)


fibonacciWithHelper :: Integer -> Integer
fibonacciWithHelper 0 = 0
fibonacciWithHelper 1 = 1
fibonacciWithHelper n | n <= 0 = helperMinus 1 0 n
                      | n > 1 = helper 1 0 n

helper acc prev 0 = prev
helper acc prev next = helper prev (acc + prev) (next - 1)

helperMinus acc prev 0 = prev
helperMinus acc prev next = helperMinus prev (acc - prev) (next + 1)

File: fibonacci.js
const fibonacci = (n) => {
    if (n === 0) return 0
    if (n === 1) return 1
    let acc = 1
    let prev = 0
    let temp
    while (n > 1) {
        temp = acc;
        acc = prev + acc;
        prev = temp;
        n--;
    }
    return acc
}

const fibonacciSlow = (n) => {
    if (n === 0) return 0
    if (n === 1) return 1
    return fibonacciSlow(n-1) + fibonacciSlow(n-2)
}

const result = fibonacci(41)
const resultSlow = fibonacciSlow(41)
console.log(result)
console.log(resultSlow)
